kratos/utilities/timer.cpp

```diff
@@ -0,0 +1,122 @@
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Pooyan Dadvand
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/timer.h"
+#include "input_output/logger.h"
+
+namespace Kratos
+{
+/// Default constructor.
+Timer::Timer(){}
+
+void Timer::Start(std::string const& rIntervalName)
+{
+    msTimeTable[rIntervalName].SetStartTime(GetTime());
+}
+
+void Timer::Stop(std::string const& rIntervalName)
+{
+    const double stop_time = GetTime();
+    ContainerType::iterator it_time_data = msTimeTable.find(rIntervalName);
+
+    if(it_time_data == msTimeTable.end())
+        return;
+
+    it_time_data->second.Update(stop_time);
+
+    PrintIntervalInformation(rIntervalName, it_time_data->second.GetStartTime(), stop_time);
+}
+
+int Timer::SetOuputFile(std::string const& rOutputFileName)
+{
+    if(msOutputFile.is_open())
+        msOutputFile.close();
+
+    msOutputFile.open(rOutputFileName.c_str());
+
+    msOutputFile << "                                         Start   \tStop     \tElapsed " << std::endl;
+
+    return msOutputFile.is_open();
+}
+
+int Timer::CloseOuputFile()
+{
+    if(msOutputFile.is_open())
+        msOutputFile.close();
+
+    return msOutputFile.is_open();
+}
+
+bool Timer::GetPrintOnScreen()
+{
+    return msPrintOnScreen;
+}
+
+void Timer::SetPrintOnScreen(bool const PrintOnScreen)
+{
+    msPrintOnScreen = PrintOnScreen;
+}
+
+void Timer::PrintIntervalInformation(std::string const& rIntervalName, const double StartTime, const double StopTime)
+{
+    if (msOutputFile.is_open()) {
+        msOutputFile << rIntervalName << " ";
+
+        for(int i = rIntervalName.size() + 1 ; i < 40 ; i++)
+            msOutputFile << ".";
+
+        msOutputFile << " " << StartTime << "s     \t" << StopTime << "s     \t" << StopTime - StartTime <<"s" << std::endl;
+    } else if(msPrintOnScreen) {
+        KRATOS_INFO("Timer") << rIntervalName << " ";
+
+        for(int i = rIntervalName.size() + 1 ; i < 40 ; i++)
+            KRATOS_INFO("Timer") << ".";
+
+        KRATOS_INFO("Timer") << " " << StartTime << "s     \t" << StopTime << "s     \t" << StopTime - StartTime <<"s" << std::endl;
+    }
+}
+
+void Timer::PrintTimingInformation()
+{
+    if(msOutputFile.is_open())
+        PrintTimingInformation(msOutputFile);
+    else if(msPrintOnScreen)
+        PrintTimingInformation(std::cout);
+}
+
+void Timer::PrintTimingInformation(std::ostream& rOStream)
+{
+    const double global_elapsed_time = ElapsedSeconds(mStartTime);
+    rOStream << "                                 Repeat # \tTotal     \tMax     \tMin     \tAverage     \t%" << std::endl;
+    for(auto& r_time_data : msTimeTable) {
+        rOStream << r_time_data.first;
+        for(int i =  r_time_data.first.size() + 1 ; i < 40 ; i++)
+            rOStream << ".";
+
+        rOStream << " ";
+        r_time_data.second.PrintData(rOStream, global_elapsed_time);
```

Comment:
Can we add format here to `global_elapse_time`? So that all outputs have something like 5 decimals or so.

Different lengths of the timing screw up the format of the final table (see SolverInitialize):

```
Build.................................. 17 	0.227601s     	0.0146794s     	0.0121621s     	0.0133883s     	3.49373%
InitializeSolutionStep................. 5 	0.313942s     	0.0746375s     	0.0591535s     	0.0627885s     	4.81908%
MatrixStructure........................ 5 	0.146374s     	0.0393995s     	0.025908s     	0.0292748s     	2.24688%
Reading Input.......................... 1 	0.217838s     	0.217838s     	0.217838s     	0.217838s     	3.34387%
Solve.................................. 17 	2.73678s     	0.568448s     	0.117503s     	0.160987s     	42.0102%
SolveSolutionStep...................... 5 	3.05012s     	1.26766s     	0.435465s     	0.610024s     	46.82%
SolverInitialize....................... 1 	0.000392116s     	0.000392116s     	0.000392116s     	0.000392116s     	0.00601907%
Writing Mesh........................... 1 	0.0425325s     	0.0425325s     	0.0425325s     	0.0425325s     	0.652883%
Writing Results........................ 10 	0.0472715s     	0.00690678s     	0.0025723s     	0.00472715s     	0.725628%

```

Reply:
Also, it seems the final timings are sorted alphabetically. Is it viable to sort them by order of call?

Reply:
The structure is a map, not unordered map, the order is according to name